### PR TITLE
refactor(connlib): implement new FFI guidelines

### PR DESCRIPTION
**Pull request by thomaseizinger**
This updates connlib to follow the new guidelines described in #4262. I only made the bare-minimum changes to the clients. With these changes `reconnect` should only be called when the network interface actually changed, meaning clients have to be updated to reflect that.

### CODE REVIEW

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [inline rust/connlib/clients/android/src/lib.rs:536]: This is good -- I think we'll see this in Firebase.

File: rust/connlib/clients/shared/src/lib.rs
Comment (jamilbk) [inline rust/connlib/clients/shared/src/lib.rs:76]:
```suggestion
    /// 1. On MacOS, a socket bound to the unspecified IP cannot send to interfaces attached after the socket has been created.
```

File: rust/connlib/clients/shared/src/lib.rs
Comment (jamilbk) [inline rust/connlib/clients/shared/src/lib.rs:77]:
```suggestion
    /// 2. Switching between networks or network interfaces changes the 3-tuple of the client.
```

**Review by jamilbk — APPROVED** [review]
LGTM! I'll rebase #4133 after this is merged

**Comment by jamilbk** [comment]
@conectado @ReactorScram --

I believe with this PR you will want to implement similar logic as I did in #4133 in order to be smarter on the client side about calling `reconnect` and `set_dns`:

- Only call `reconnect` if the primary interface has changed (or default egress interface)
- Only call `set_dns` if the servers have changed

File: rust/connlib/tunnel/src/lib.rs
Comment (ReactorScram) [inline rust/connlib/tunnel/src/lib.rs:255]: If it's possible to protect the file descriptors from inside connlib and inside the tunnel process, why did it move outside? Is it for error handling?

This is something that's confused me a little. Connlib _can_ be platform-specific, right? Like `tun_linux.rs` and `tun_windows.rs`. We're not just moving things out to clients and across the FFI boundary just because, right?

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/gui.rs:537]: That #2986 would be really nice here for a default.

**Review by ReactorScram — APPROVED** [review]
It looks okay but I do not understand what the advantage is.

File: rust/connlib/clients/shared/src/lib.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/lib.rs:93]: If we're going to test this from the client-side we don't need this here, in which case we don't need to keep track of the system's resolver in connlib.

**Review by conectado — APPROVED** [review]

File: rust/connlib/tunnel/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs:255]: In this case, `protect` can only be called from JVM code so we need a callback.

I think platform-specific callbacks are confusing. I'd rather shuffle the logic around so the individual client glue-code crates can do this.

I'd still call this "connlib being platform specific" because the Android app just uses a `Session`.

One alternative could be to store a dynamic-dispatched callback pointer inside `Sockets`, maybe that is even cleaner? It would be "invisible" from the outside.

File: rust/gui-client/src-tauri/src/client/gui.rs
Comment (thomaseizinger) [reply rust/gui-client/src-tauri/src/client/gui.rs:537]: Great minds think alike, that is why I reopened it 😁

File: rust/connlib/clients/shared/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:93]: Do you think it is worth getting rid of that state tracking? I am a bit torn. It seems easy to make it idempotent which means a little less code on the clients?

Happy to go either way here.

Comment (jamilbk) [reply rust/connlib/clients/shared/src/lib.rs:93]: > don't need to keep track of the system's resolver in connlib.

At all? Or just the last few. We'll still have the `set_dns` right?

File: rust/connlib/tunnel/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs:255]: What do you think of [2c2c617](https://github.com/firezone/firezone/pull/4263/commits/2c2c6178cbfa9b5d6ed7ee69aef3727ba6ebced2)?

File: rust/connlib/clients/shared/src/lib.rs
Comment (conectado) [reply rust/connlib/clients/shared/src/lib.rs:93]: >At all? Or just the last few. We'll still have the set_dns right?

we would still have it, but we only store the `system_resolvers` in connlib to make `set_dns` not react to the same servers

Comment (conectado) [reply rust/connlib/clients/shared/src/lib.rs:93]: > Do you think it is worth getting rid of that state tracking? I am a bit torn. It seems easy to make it idempotent which means a little less code on the clients?

I just think that we shouldn't keep the same logic in two places and since it seems we are going with the route of having this logic in the client I'd just remove it from connlib.

**Comment by conectado** [comment]
rebased to solve conflicts

**Comment by jamilbk** [comment]
Unfortunately this PR is now crashing the Apple clients again... :-(

Edit: Not completely sure it was this PR. The working commit I was using on #4133 was fa61a84ee, here's the diff after rebasing

https://github.com/firezone/firezone/compare/fa61a84ee..fe239a954da2e09b68e53d8e08d640421b479f37

**Comment by thomaseizinger** [comment]
@conectado Was there a reason why you dropped https://github.com/firezone/firezone/pull/4263/commits/2c2c6178cbfa9b5d6ed7ee69aef3727ba6ebced2 from this PR?

**Comment by conectado** [comment]
> @conectado Was there a reason why you dropped [2c2c617](https://github.com/firezone/firezone/commit/2c2c6178cbfa9b5d6ed7ee69aef3727ba6ebced2) from this PR?

@thomaseizinger Not on pupose, seems to have been a mistake when I rebased

**Pull request merged**
